### PR TITLE
RoR2: Remove recursion from explore mode access rules

### PR DESCRIPTION
**Pull request by Mysteryem**
## What is this fixing or adding?

The access rules for "\<Environment name> Chest n", "\<Environment name> Shrine n" etc. locations recursively called state.can_reach() for the n-1 location name, with the n=1 location being the only location to have the actual access rule set.

This patch removes the recursion, instead setting the actual access rule directly on each location, increasing the performance of checking accessibility of n>1 locations.

## How was this tested?

Risk of Rain 2 was already quite fast to generate despite the recursion in the access rules, but with this patch, generating a multiworld with 200 copies of the template RoR2 yaml (and progression balancing disabled through a meta.yaml) goes from about 18s to about 6s for me.

From generating the same seed before and after this patch, the same result is produced.

### CODE REVIEW

**Comment by Mysteryem** [comment]
@kindasneaki since you are the RoR2 maintainer

**Review by kindasneaki — APPROVED** [review]
Looks good to me. I did a couple generations and looked at the logic to make sure everything looked normal. I also noticed a speed increase in generation time.

**Review by Exempt-Medic — APPROVED** [review]
The code LGTM, generations before and after with 100 fully-random yamls yielded identical results with a ~62.% reduction in generation time

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**
